src/status_im2/contexts/chat/photo_selector/view.cljs

```diff
@@ -0,0 +1,108 @@
+            [react-native.linear-gradient :as linear-gradient]
+            [reagent.core :as reagent]
+            [status-im2.contexts.chat.photo-selector.style :as style]
+            [status-im.utils.core :as utils]
+            [utils.re-frame :as rf]))
+
+(def selected (reagent/atom []))
+
+(defn bottom-gradient
+  [chat-id selected-images]
+  [:f>
+   (fn []
+     (let [safe-area (safe-area/use-safe-area)]
+       (when (or (pos? (count @selected)) selected-images)
+         [linear-gradient/linear-gradient
+          {:colors [:black :transparent]
+           :start  {:x 0 :y 1}
+           :end    {:x 0 :y 0}
+           :style  (style/gradient-container safe-area)}
+          [quo2/button
+           {:style               {:align-self        :stretch
+                                  :margin-horizontal 20}
+            :on-press            #(do
+                                    (rf/dispatch [:chat.ui/clear-sending-images chat-id])
+                                    (doseq [item @selected]
+                                      (rf/dispatch [:chat.ui/camera-roll-pick item]))
+                                    (reset! selected [])
+                                    (rf/dispatch [:bottom-sheet/hide]))
+            :accessibility-label :confirm-selection}
+           (i18n/label :t/confirm-selection)]])))])
+
+(defn clear-button
+  []
+  (when (pos? (count @selected))
+    [rn/touchable-opacity
+     {:on-press            #(reset! selected [])
+      :style               (style/clear-container)
+      :accessibility-label :clear}
+     [quo2/text {:weight :medium} (i18n/label :t/clear)]]))
+
+(defn image
+  [item index _ {:keys [window-width]}]
+  [rn/touchable-opacity
+   {:active-opacity      1
+    :on-press            (fn []
+                           (if (some #{item} @selected)
+                             (reset! selected (vec (remove #(= % item) @selected)))
+                             (swap! selected conj item)))
+    :accessibility-label (str "image-" index)}
+   [rn/image
+    {:source {:uri item}
+     :style  (style/image window-width index)}]
+   (when (some #{item} @selected)
+     [rn/view {:style (style/overlay window-width)}])
+   (when (some #{item} @selected)
+     [info-count/info-count
+      {:style               style/image-count
+       :accessibility-label (str "count-" index)}
+      (inc (utils/first-index #(= item %) @selected))])])
+
+(defn photo-selector
+  [chat-id]
+  (rf/dispatch [:chat.ui/camera-roll-get-photos 20])
+  (let [selected-images (keys (get-in (rf/sub [:chat/inputs]) [chat-id :metadata :sending-image]))]
+    (when selected-images
+      (reset! selected (vec selected-images)))
+    [:f>
+     (fn []
+       (let [{window-height :height window-width :width} (rn/use-window-dimensions)
+             safe-area                                   (safe-area/use-safe-area)
+             camera-roll-photos                          (rf/sub [:camera-roll/photos])
```

Comment:
No action on your part: FYI: I know we've discussed this topic in the past, but just a reminder, these top level subscriptions impact re-frame's performance because they are root subscriptions and need to be recomputed on *every* state change.

Reply:
Is `(rf/sub [:camera-roll/XYZ])` a root sub?

Reply:
The subscription `:camera-roll/XYZ` is just a single keyword. We could transform it into `:camera-roll.iphone.14/x/y/z` and it would still be a single keyword. Hence there's no relationship between the keyword name and how nested a map is. Namespaced keywords do follow a convention of using dots to convey hierarchy, and maybe that's one source of confusion. Dots in Clojure keywords mean nothing, except convention in certain cases. Nested maps are never accessed using dots in Clojure, we need `get`, `get-in`, etc.

So in the code we have:

```clojure
(reg-root-key-sub :camera-roll/photos :camera-roll/photos)
```

So in the app DB it's stored at the root, as `{:camera-roll/photos ...}`, not `{:camera-roll {:photos ...}}`.

In order to reduce the performance penalty I mentioned, we would need to redesign the app state to register a root subscription called `:camera-roll`, and then we would register layer-2 and layer-3 subscriptions.

I hope the explanation helps :)

I